sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -807,3 +809,37 @@ int64_t CurlSession::ResponseBufferParser::BuildHeader(
   // Parsing Headers will make sure to move one possition
   return indexOfEndOfStatusLine + 1 - buffer;
 }
+
+std::list<std::unique_ptr<CurlSession::CurlConnection>> CurlSession::c_connectionPool;
```

Comment:
Is this global variable? So the connections will never be closed and customer has no way to close connections?

Reply:
will it work if application connects to multiple hosts at the same time?

Reply:
Do we need mutex to protect this pool?

Reply:
connection is clean up on the CurlConnection destructor.
There is code to clean expired connections from the list when a new connection is requested.

Or, when application is terminated, any connection in the pool would be clean up (destructor called)

Reply:
yes, Mutex is a TODO.

and yes, each connection is related to a host.
Some expected use cases would be:
-  Multiple threads creating a connection with same pipeline to same host. In this case, connection can't be re-used. each thread would create his own connection.
-  After some time, some other thread can reuse any connection available in the pool.
-  Connections get expired after some inactivity time. Pool gets clean and connections are clean up every time a connection is requested.

Reply:
> connection is clean up on the CurlConnection destructor.
> There is code to clean expired connections from the list when a new connection is requested.
> 
> Or, when application is terminated, any connection in the pool would be clean up (destructor called)

What if there are no new connection is requested? e.g. we start 30k requests at the same time. Then these 30k connections will be alive all the time until the program exits. Our customers may not want their program to exit. Maybe it runs as a service, for a long time.

Reply:
I was thinking we can store this pool in `TransportPolicy`. So that a container client and all blob clients that propagate from this container client share the same policy via shared pointer and thus the same pool. These contain and blob clients almost always send requests to the same host, so we can make the best use of connection keep-alive.

When the last of of these client gets destructed, the policy is also destructed and the connections in the pool are also released. This way, our customer can better control when the connections can be released.

Reply:
BTW, is it defined behavior if we have this global variable and this library is compiled as a dll on Windows? @CaseyCarter 
@BillyONeal 

I remember things we can do during dll initialization are very limited.

Reply:
This should not be a problem if we compile to a DLL.